File: src/lib/prism-axis-compatibility.ts
const PRISM_AXIS_COMPONENT_TYPES = ["subjective", "final-subjective"] as const;

const HORIZONTAL_ALIASES = [
  ["r_pris", "r_pr_h"],
  ["l_pris", "l_pr_h"],
  ["r_base", "r_base_h"],
  ["l_base", "l_base_h"],
] as const;

export const PRISM_AXIS_VERTICAL_FIELDS = [
  "r_pr_v",
  "l_pr_v",
  "r_base_v",
  "l_base_v",
] as const;

const EYE_HORIZONTAL_LINKS: Record<string, string> = {
  pris: "pr_h",
  pr_h: "pris",
  base: "base_h",
  base_h: "base",
};

const isRecord = (value: unknown): value is Record<string, unknown> =>
  Boolean(value) && typeof value === "object" && !Array.isArray(value);

const isPresent = (value: unknown) =>
  value !== undefined && value !== null && String(value).trim() !== "";

const isFilled = (value: unknown) => {
  if (!isPresent(value)) return false;
  const trimmed = String(value).trim();
  return trimmed !== "0" && trimmed !== "0.0";
};

export class PrismAxisCompatibility {
  static isComponentKey(key: string) {
    return PRISM_AXIS_COMPONENT_TYPES.some(
      (type) => key === type || key.startsWith(`${type}-`),
    );
  }

  static hasVerticalPrism(data: Record<string, unknown> | null | undefined) {
    if (!data) return false;
    return PRISM_AXIS_VERTICAL_FIELDS.some((field) => isFilled(data[field]));
  }

  static normalizeBlock<T extends Record<string, unknown>>(data: T): T {
    const next: Record<string, unknown> = { ...data };
    HORIZONTAL_ALIASES.forEach(([legacy, split]) => {
      const legacyPresent = isPresent(next[legacy]);
      const splitPresent = isPresent(next[split]);
      if (legacyPresent) {
        next[split] = next[legacy];
      } else if (splitPresent) {
        next[legacy] = next[split];
      }
    });
    return next as T;
  }

  static normalizeExamData<T extends Record<string, unknown>>(data: T): T {
    const normalized: Record<string, unknown> = {};
    Object.entries(data).forEach(([key, value]) => {
      if (this.isComponentKey(key) && isRecord(value)) {
        normalized[key] = this.normalizeBlock(value);
        return;
      }
      normalized[key] = value;
    });
    return normalized as T;
  }

  static linkedHorizontalField(field: string) {
    return EYE_HORIZONTAL_LINKS[field];
  }
}

File: backend/services/prism_axis_compatibility.py
from typing import Any, Dict, Mapping

PRISM_AXIS_COMPONENT_TYPES = ("subjective", "final-subjective")

HORIZONTAL_ALIASES = (
    ("r_pris", "r_pr_h"),
    ("l_pris", "l_pr_h"),
    ("r_base", "r_base_h"),
    ("l_base", "l_base_h"),
)


def _is_present(value: Any) -> bool:
    if value is None:
        return False
    if isinstance(value, str) and not value.strip():
        return False
    return True


def is_prism_axis_component_key(key: str) -> bool:
    return any(
        key == component_type or key.startswith(f"{component_type}-")
        for component_type in PRISM_AXIS_COMPONENT_TYPES
    )


def normalize_prism_axis_block(block: Mapping[str, Any]) -> Dict[str, Any]:
    next_block = dict(block)
    for legacy, split in HORIZONTAL_ALIASES:
        legacy_present = _is_present(next_block.get(legacy))
        split_present = _is_present(next_block.get(split))
        if legacy_present:
            next_block[split] = next_block.get(legacy)
        elif split_present:
            next_block[legacy] = next_block.get(split)
    return next_block


def normalize_prism_axis_exam_data(exam_data: Dict[str, Any] | None) -> Dict[str, Any]:
    if not exam_data:
        return {}
    normalized: Dict[str, Any] = {}
    for key, value in exam_data.items():
        if is_prism_axis_component_key(key) and isinstance(value, dict):
            normalized[key] = normalize_prism_axis_block(value)
            continue
        normalized[key] = value
    return normalized
